numba/tests/test_stencils.py

```diff
@@ -365,12 +407,1272 @@ def test_impl(n):
+    # or not.
+    def test_basic31b(self):
+        """does a const propagate?"""
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            s = 1
+            t = 1 - s
+            return a[t, 0]
+        self.check(kernel, a)
+
+    def test_basic32(self):
+        """typed int index"""
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            return a[np.int8(1), 0]
+        self.check(kernel, a, expected_exception=[ValueError, LoweringError])
+
+    def test_basic33(self):
+        """add 0d array"""
+        a = np.arange(12.).reshape(3, 4)
+
+        def kernel(a):
+            return a[0, 0] + np.array(1)
+        self.check(kernel, a)
+
+    def test_basic34(self):
+        """More complex rel index with dependency on addition rel index"""
+        def kernel(a):
+            g = 4. + a[0, 1]
+            return g + (a[0, 1] + a[1, 0] + a[0, -1] + np.sin(a[-2, 0]))
+        a = np.arange(144).reshape(12, 12)
+        self.check(kernel, a)
+
+    # parfors ought to raise but does not, is this likely down to the allocation of the
+    # return array under parfors being a different type?
+    def test_basic35(self):
+        """simple cval """
+        def kernel(a):
+            return a[0, 1]
+        a = np.arange(12.).reshape(3, 4)
+        ex = self.exception_dict(stencil=ValueError, parfor=LoweringError, njit=LoweringError)
+        self.check(kernel, a, options={'cval': 5}, expected_exception=ex)
+
+    def test_basic36(self):
+        """more complex with cval"""
+        def kernel(a):
+            return a[0, 1] + a[0, -1] + a[1, -1] + a[1, -1]
+        a = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, options={'cval': 5.})
+
+    def test_basic37(self):
+        """cval is expr"""
+        def kernel(a):
+            return a[0, 1] + a[0, -1] + a[1, -1] + a[1, -1]
+        a = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, options={'cval': 5 + 63.})
+
+    # pyStencil handles the type change, the others raise
+    def test_basic38(self):
+        """cval is complex"""
+        def kernel(a):
+            return a[0, 1] + a[0, -1] + a[1, -1] + a[1, -1]
+        a = np.arange(12.).reshape(3, 4)
+        ex = self.exception_dict(stencil=ValueError, parfor=TypeError, njit=LoweringError)
+        self.check(kernel, a, options={'cval': 1.j}, expected_exception=ex)
+
+    def test_basic39(self):
+        """cval is func expr"""
+        def kernel(a):
+            return a[0, 1] + a[0, -1] + a[1, -1] + a[1, -1]
+        a = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, options={'cval': np.sin(3.) + np.cos(2)})
+
+    def test_basic40(self):
+        """2 args!"""
+        def kernel(a, b):
+            return a[0, 1] + b[0, -2]
+        a = np.arange(12.).reshape(3, 4)
+        b = np.arange(12.).reshape(3, 4)
+        self.check(kernel, a, b)
+
+    # This fails pyStencil but stencil func runs, fairly sure it shouldn't?
+    def test_basic41(self):
```

Comment:
This caused a lot of confusion. I *think* `b` is being accessed out of bounds and the correct behaviour should be asserting that `a` and `b` have the same shape?

Reply:
We are leaning towards an equality check now.  We are currently calling a function that does the check for Python and njit modes.  We will be adding an assertion in ParallelAccelerator through array analysis that all the relative indexed arrays are the same size.  The first part is done.  The ParallelAccelerator part will take a few days.

Reply:
Okay.  We are generating an assertion for this now on the PA side.

Reply:
That's great thanks.